src/include/concepts.h

```diff
@@ -36,78 +36,196 @@
+template <class T>
+concept partitionable = requires(const T& t) {
+  //  { num_partitions(t) } -> semi_integral;
+  { t.indices() };
+  { t.ids() };
+};
+
+// ----------------------------------------------------------------------------
+// feature_vector concept
+// ----------------------------------------------------------------------------
+
+template <typename R>
+concept feature_vector =
+    std::ranges::random_access_range<R> && /* std::ranges::sized_range<R> && */
+    std::ranges::contiguous_range<R> && dimensionable<R> &&
+    (subscriptable_range<R> || callable_range<R>);
+
+template <class R>
+concept query_vector = feature_vector<R>;
+
+// ----------------------------------------------------------------------------
+// feature_vector_array concept
+//
+// A vector range is a range of feature vectors.
+// We will want at least these kinds:
+//   contiguous -- all the vectors are stored in a single contiguous array
+//   partitioned -- the range is a range of partitions, each of which is a
+//                  contiguous range of vectors
+//   distributed?-- the range is a range of partitions, each of which is a
+//
+// @todo operator()(size_t, size_t) ?
+// ----------------------------------------------------------------------------
+template <class D>
+concept feature_vector_array =
+    // feature_vector<inner_range_t<D>> &&
+    // std::ranges::random_access_range<D> && /* std::ranges::sized_range<D> &&
+    // */ subscriptable_range<D> && requires(D d, const
+    // std::iter_difference_t<std::ranges::iterator_t<D>> n) {
+    requires(D d, size_t n) {
+      //      typename D::feature_type;
+      { num_vectors(d) } -> semi_integral;
+      { dimension(d) } -> semi_integral;
+      //{
+      //        d[n]
+      //      } ->
+      //      std::same_as<std::iter_reference_t<std::ranges::iterator_t<D>>>;
+      { d[n] } -> feature_vector;  // Maybe redundant
+    };
+
+/**
+ * @brief A concept for contiguous vector ranges.  The member function data()
+ * returns a pointer to the underlying contiguous one-dimensional storage.
+ * @tparam D
+ *
+ */
+// @todo -- add ranges::contiguous_range as a requirement
+template <class D>
+concept contiguous_feature_vector_array =
+    feature_vector_array<D> && requires(D d) {
+      { data(d) } -> std::same_as<std::add_pointer_t<typename D::reference>>;
+    };
+
+template <class T>
+concept query_vector_array = feature_vector_array<T>;
+
+template <class T>
+concept contiguous_query_vector_array = contiguous_feature_vector_array<T>;
+
+// ----------------------------------------------------------------------------
+// partitioned_feature_vector_range concept (WIP)
+// ----------------------------------------------------------------------------
+
+template <class D>
+concept partitioned_feature_vector_array =
+    feature_vector_array<D> && partitionable<D>;
+//    && requires(D d) {
+//      typename D::id_type;
+//      typename D::indices_type;
+//    };
```

Comment:
(Minor) Remove, please. Will stop reviewing for this, but let's remove all commented out code please.